### DISCUSSION TITLE
UseAfter or afterRoutesInits

### DISCUSSION THREAD

**Discussion by zantler** [TypedProject/ts-express-decorators#1611] [Q&A] [answered]
Hi Its me again!

Im strugging running a global middleware after an enpoints end.... yes it needs to be global due the ammount of apis we handle.

On the server.ts i put
`$afterRoutesInit(): void {
    this.use(TestMiddle);
  }`

where TestMiddle is the Middleware that consolo log a hello world, but its never called. 

Can you help me to figure out how to use it correctly?

Thnks a lot

**Comment by Romakita** [comment, marked as answer]
Hello @zantler 

This is the base of express.js and middlewares ^^. Unfortunatly, the chance to execute the middleware at the end of the list is low. The middlewares are only execute until that the response isn’t sent to your consummer. 

So middlewares added via $afterRoutesInit hook are pushed after all routes and it’s generally to late to do something. Especially if you want to log something.

**Reply by zantler** [reply]
Well i need to retrieve some data from the request and the response.... im trying to understand how the LogIncomingRequestMiddleware is called after the endpoint resolves but its not to clear for me and i need a similar slution to my middleware

**Reply by Romakita** [reply]
LogIncomingRequestMiddleware is a global middleware added before all routes and listen a special request (or response) event. 

But it haven’t knowledge about the called endpoint, because a single request can call one or more endpoints methods. 

The actual LogRequest is able to map request/response info to the log. But if you have to log specific things depending on the called endpoint, it will be complicated. 

I need more details ;)

**Reply by zantler** [reply]
Well... i need a lot of info for the request and the statusCode- Status mesage and the outputsize of and endpoit. It in the usebefore ir works but the info for the response is null... i dont know if it is a good explain :/

**Reply by Romakita** [reply]
The response is null until the request isn’t fullfilled ^^
But I see. I’m not on my computer so is complication for me to give you a good example. I’ll try to give you more details asap.

**Reply by zantler** [reply]
bro you are awesome, i'll wait then!

**Reply by Romakita** [reply]
Here is the example:

```typescript
@OverrideProvide(LogIncomingRequestMiddleware)
export class CustomLogIncomingRequestMiddleware extends LogIncomingRequestMiddleware {
 protected onLogEnd(request: Req, response: Res) {
    const {debug, logRequest, logEnd} = this.settings;

    if (logEnd !== false) {
      if (debug) {
        request.ctx.logger.debug({
          event: "request.end",
          status: response.statusCode,
          data: request.ctx.data
        });
      } else if (logRequest) {
        request.ctx.logger.info({
          event: "request.end",
          status: response.statusCode,
          // add here the response size,
          size: response.get('Content-size') // or a method to calculate the response size from request.ctx.data
          // then add extra needed data
        });
      }
    }

    request.ctx.logger.flush();
  }
```

**Reply by zantler** [reply]
thaks bro!! that helps!!
